odoo/fields.py

```diff
@@ -1398,10 +1398,17 @@ def _setup_related_full(self, model):
 
     def convert_to_column(self, value, record, values=None, validate=True):
         # retrieve currency from values or record
+        c_field = record._fields[self.currency_field]
         if values and self.currency_field in values:
-            field = record._fields[self.currency_field]
-            currency = field.convert_to_cache(values[self.currency_field], record, validate)
-            currency = field.convert_to_record(currency, record)
+            currency = c_field.convert_to_cache(values[self.currency_field], record, validate)
+            currency = c_field.convert_to_record(currency, record)
+        elif values and c_field.related and c_field.related[0] in values:
+            # c_field.related[0] is a relational field, use convert_to_record to return an appropriated recordset
+            relation_field = record._fields[c_field.related[0]]
+            currency = relation_field.convert_to_cache(values[c_field.related[0]], record, validate)
+            currency = relation_field.convert_to_record(currency, record)[:1]
+            for field in c_field.related[1:]:
+                currency = currency[field][:1]
```

Comment:
Why don't you use a new record instead?  This looks simpler and more robust:
```suggestion
            rel_fname = c_field.related[0]
            dummy = record.new({rel_fname: values[rel_fname]})
            currency = dummy[self.currency_field]
```
All the conversions and computations are handled by the `dummy` record itself.
The same trick can be used in the case above, by the way.